lib/inv_ca_cg.cpp

```diff
@@ -12,37 +12,35 @@
 
 namespace quda {
 
-  enum Basis {
-    POWER_BASIS,
-    INVALID_BASIS
-  };
-
-  const static Basis basis = POWER_BASIS;
-
   CACG::CACG(DiracMatrix &mat, DiracMatrix &matSloppy, SolverParam &param, TimeProfile &profile)
     : Solver(param, profile), mat(mat), matSloppy(matSloppy), init(false),
-      W(nullptr), C(nullptr), alpha(nullptr), beta(nullptr), phi(nullptr), rp(nullptr),
+      lambda_init(false), basis(param.ca_basis),
+      Q_AQandg(nullptr), Q_AS(nullptr), alpha(nullptr), beta(nullptr), rp(nullptr),
       tmpp(nullptr), tmpp2(nullptr), tmp_sloppy(nullptr), tmp_sloppy2(nullptr) { }
 
   CACG::~CACG() {
     if (!param.is_preconditioner) profile.TPSTART(QUDA_PROFILE_FREE);
 
     if (init) {
-      if (W) delete []W;
-      if (C) delete []C;
+      if (Q_AQandg) delete []Q_AQandg;
+      if (Q_AS) delete []Q_AS;
       if (alpha) delete []alpha;
       if (beta) delete []beta;
-      if (phi) delete []phi;
-      bool use_source = (param.preserve_source == QUDA_PRESERVE_SOURCE_NO &&
-                         param.precision == param.precision_sloppy &&
-                         param.use_init_guess == QUDA_USE_INIT_GUESS_NO);
-      if (basis == POWER_BASIS) {
-        for (int i=0; i<param.Nkrylov+1; i++) if (i>0 || !use_source) delete r[i];
+
+      bool use_source = false; // needed for explicit residual recompute
+                         // (param.preserve_source == QUDA_PRESERVE_SOURCE_NO &&
+                         // param.precision == param.precision_sloppy &&
+                         // param.use_init_guess == QUDA_USE_INIT_GUESS_NO);
+      if (basis == QUDA_POWER_BASIS) {
+        for (int i=0; i<param.Nkrylov+1; i++) if (i>0 || !use_source) delete S[i];
       } else {
-        for (int i=0; i<param.Nkrylov; i++) if (i>0 || !use_source) delete r[i];
-        for (int i=0; i<param.Nkrylov; i++) delete q[i];
+        for (int i=0; i<param.Nkrylov; i++) if (i>0 || !use_source) delete S[i];
+        for (int i=0; i<param.Nkrylov; i++) delete AS[i];
       }
-      for (int i=0; i<param.Nkrylov; i++) delete p[i];
+      for (int i=0; i<param.Nkrylov; i++) delete Q[i];
+      for (int i=0; i<param.Nkrylov; i++) delete Qtmp[i];
+      for (int i=0; i<param.Nkrylov; i++) delete AQ[i];
```

Comment:
clang-format suggested the following change:


```
-      for (int i=0; i<param.Nkrylov; i++) delete Q[i];
-      for (int i=0; i<param.Nkrylov; i++) delete Qtmp[i];
-      for (int i=0; i<param.Nkrylov; i++) delete AQ[i];
+      for (int i = 0; i < param.Nkrylov; i++) delete Q[i];
+      for (int i = 0; i < param.Nkrylov; i++) delete Qtmp[i];
+      for (int i = 0; i < param.Nkrylov; i++) delete AQ[i];
```

<!-- identifier: clang-format -->
<!-- File: lib/inv_ca_cg.cpp -->